### PR TITLE
fix-#159 : Configured Theme changes to light mode on page refresh in add blogs page

### LINKED ISSUE #159: [BUG] Theme changes to light mode on page refresh in add blogs page
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When the user has a dark theme selected and refreshes the page, the theme changes to light mode.

https://github.com/krishnaacharyaa/wanderlust/assets/46074384/1f95ad4c-189b-4d3d-865d-c8fa3443f778

### Expected Behavior

It should not switch to light theme on page refresh.

### Reproduction steps

```bash
+ Select dark mode
+ Refresh the page
```

### Browsers

Chrome

### OS

Windows

### PR DESCRIPTION
**Pull request by Ritik-The-Dev**
## Summary

Fixed the issue of the theme changing to light mode on page refresh in the add blogs page by adding the class "dark" to the document if the theme is set to "dark" in local storage.

## Description

The issue occurred because the class list did not include "dark" while in dark mode, so I added it when the theme is set to "dark" in local storage.

## Issue(s) Addressed
Closes #159 

## Prerequisites

- [x] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
Hi @Ritik-The-Dev thank you for the PR, please create a new PR which follow proper commit guidelines. and take care of issues addressed section in PR description section

But we are looking for some better way to handle this tbh, because we can't have these lines everywhere in all the pages, can you kindly dig deep and figure out what is going wrong, We are expecting some core logic change and it should server to all the pages from one place

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -9,8 +9,15 @@ import SignUp from '@/pages/signup-page';
 import AdminUsers from '@/pages/admin-users';
 import AdminBlogs from '@/pages/admin-blogs';
 import AdminContainer from './components/admin-container';
+import { useEffect } from 'react';
 
 function App() {
+
+  useEffect(() => {
+    const storedTheme = localStorage.getItem('theme');
```
Comment (krishnaacharyaa) [inline frontend/src/App.tsx:17]: useLayoutEffect may be better? Kindly see which would be more relevant. And can you even evaluate and optimize the toggle-button file, if there is scope? I feel it has some uneccessary code. 

And it is better to have this logic somewhere else and import in the main file than here, helps in seperation of concerns. Please do some research and get back, happy to know your inputs

**Review by krishnaacharyaa — COMMENTED** [review]
Good fix, thank you
Request for optimization even further, kindly address

File: frontend/src/App.tsx (hunk 1 above)
Comment (Ritik-The-Dev) [reply frontend/src/App.tsx:17]: Ok Sir

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -9,8 +9,13 @@ import SignUp from '@/pages/signup-page';
 import AdminUsers from '@/pages/admin-users';
 import AdminBlogs from '@/pages/admin-blogs';
 import AdminContainer from './components/admin-container';
+import { useEffect } from 'react';
+import useThemeClass from './utils/theme-changer';
 
 function App() {
+  useEffect(() => {
+    useThemeClass();
+  }, []);
```
Comment (krishnaacharyaa) [inline frontend/src/App.tsx:18]: useLayoutEffect maybe?

File: frontend/src/components/theme-toggle-button.tsx [hunk 3]
```diff
@@ -23,25 +23,20 @@ function ThemeToggle() {
...
-      <button
-        name="theme"
-        className={`${
-          isDarkTheme ? 'bg-dark-theme-background' : 'bg-light-theme-background'
-        } flex h-8 w-16 cursor-pointer items-center justify-start rounded-full px-1 py-1`}
-        onClick={toggleTheme}
+    <button
+      name="theme"
+      className={`bg-${isDarkTheme ? 'dark' : 'light'}-theme-background 
+      flex h-8 w-16 cursor-pointer items-center justify-start rounded-full px-1 py-1`}
```
Comment (krishnaacharyaa) [inline frontend/src/components/theme-toggle-button.tsx:29]: Please revert this, the previous one is much readable and we cannot actually use the tailwind classes this way

File: frontend/src/components/theme-toggle-button.tsx [hunk 4]
```diff
@@ -23,25 +23,20 @@ function ThemeToggle() {
...
+      onClick={toggleTheme}
+    >
+      <div
+        className={`translate-x-${isDarkTheme ? '8' : '0'} bg-${
+          isDarkTheme ? 'dark' : 'light'
+        }-theme-foreground h-6 w-6 rounded-full bg-black px-1 py-1 duration-300`}
       >
```
Comment (krishnaacharyaa) [inline frontend/src/components/theme-toggle-button.tsx:36]: Same comment

File: frontend/src/components/theme-toggle-button.tsx
Comment (krishnaacharyaa) [inline frontend/src/components/theme-toggle-button.tsx:1]: I wanted you to check on this line 1 to 23 can we make it clean and optimize this even further

File: frontend/src/utils/theme-changer.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { useLayoutEffect } from 'react';
+
+const useThemeClass = () => {
+  useLayoutEffect(() => {
+    const storedTheme = localStorage.getItem('theme');
```
Comment (krishnaacharyaa) [inline frontend/src/utils/theme-changer.ts:5]: so basically now we are inside the useLayoutEffect and outside we have useEffect ? two nesting is it preffered? 

useEffect -> useLayoutEffect . Can we not eliminate this, and have parent be useLayoutEffect in the app.tsx

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
Good efforts, thank you, kindly address the reviews

File: frontend/src/components/theme-toggle-button.tsx (hunk 3 above)
Comment (Ritik-The-Dev) [reply frontend/src/components/theme-toggle-button.tsx]: Done Sir

File: frontend/src/components/theme-toggle-button.tsx (hunk 4 above)
Comment (Ritik-The-Dev) [reply frontend/src/components/theme-toggle-button.tsx]: Done Sir

File: frontend/src/utils/theme-changer.ts [hunk 6]
```diff
@@ -1,10 +1,6 @@
-import { useLayoutEffect } from 'react';
-
 const useThemeClass = () => {
-  useLayoutEffect(() => {
-    const storedTheme = localStorage.getItem('theme');
-    storedTheme === 'dark' && document.documentElement.classList.add('dark');
-  }, []);
+  const storedTheme = localStorage.getItem('theme');
+  storedTheme === 'dark' && document.documentElement.classList.add('dark');
 };
```
Comment (krishnaacharyaa) [inline frontend/src/utils/theme-changer.ts:4]: So now lets say i have dark mode enabled in my browser and then i visit the website for the first time. I don't think so that time it will work right? because it is just taking from the localstorage.... and this is done in the 2nd iteration, but the 1st walaa still we are missing. I guess we have to import some code from theme-toggle-button and port it here, 

And in that button, we just change the localstorage and the actual  theme right? 

Let's think a bit broader here. and try to figure out all the corner case

File: frontend/src/components/theme-toggle-button.tsx [hunk 7]
```diff
@@ -9,11 +9,13 @@ function ThemeToggle() {
   useLayoutEffect(() => {
     const storedTheme = localStorage.getItem('theme');
     const prefersDark = window.matchMedia?.('(prefers-color-scheme: dark)').matches;
-    setIsDarkTheme(storedTheme === 'dark' || (!storedTheme && prefersDark) || null);
-    if (storedTheme === 'dark' || (!storedTheme && prefersDark)) {
+    const initialTheme = storedTheme === 'dark' || (!storedTheme && prefersDark);
```
Comment (krishnaacharyaa) [inline frontend/src/components/theme-toggle-button.tsx:12]: Can you work on simplifying this, tooo much code cluttered by previous contributor, we need to make it easy to understand.

File: frontend/src/utils/theme-changer.ts (hunk 6 above)
Comment (Ritik-The-Dev) [reply frontend/src/utils/theme-changer.ts:4]: Ok sir will try to figure it out

File: frontend/src/components/theme-toggle-button.tsx (hunk 7 above)
Comment (Ritik-The-Dev) [reply frontend/src/components/theme-toggle-button.tsx:12]: ok Sir

File: frontend/src/components/theme-toggle-button.tsx [hunk 8]
```diff
@@ -7,21 +7,21 @@ function ThemeToggle() {
...
-    setIsDarkTheme(storedTheme === 'dark' || (!storedTheme && prefersDark) || null);
-    if (storedTheme === 'dark' || (!storedTheme && prefersDark)) {
-      document.documentElement.classList.add('dark');
-    }
+      const storedTheme = localStorage.getItem('theme') ?? '';
+      const prefersDark = window.matchMedia?.('(prefers-color-scheme: dark)').matches;
+      const initialTheme = storedTheme === 'dark' || (prefersDark && !storedTheme);
+      setIsDarkTheme(initialTheme);
+      if (initialTheme) document.documentElement.classList.add('dark');    
   }, []);
```
Comment (krishnaacharyaa) [inline frontend/src/components/theme-toggle-button.tsx:15]: Why do this guy need this logic? Can't theme-changer handle this?

**Comment by krishnaacharyaa** [comment]
Works all the cases, but initially if user is in dark mode and access the website he is landing into light theme.

To reproduce, set your appearence in chrom to dark mode and then remove the theme variable from the webiste from the console - storage. and refresh, 
This time ideally he should get dark theme, but he gets light, interstingly light theme is set though the user's selected browser setting was dark.
Can you evaluate on this? @Ritik-The-Dev and please resolve merge conflict

File: frontend/src/App.tsx [hunk 9]
```diff
@@ -33,7 +37,6 @@ function App() {
               <Route path="blogs" element={<AdminBlogs />} />
             </Route>
           </Route>
-          <Route path="*" element={<NotFound />} />
         </Routes>
```
Comment (krishnaacharyaa) [inline frontend/src/App.tsx:40]: Why is this removed? Can you please add it back

File: frontend/src/utils/theme-changer.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+const useThemeClass = () => {
+  const storedTheme = localStorage.getItem('theme');
+  const prefersDark = window.matchMedia?.('(prefers-color-scheme: dark)').matches;
+  const initialTheme = storedTheme === 'dark' || (!storedTheme && prefersDark);
+  initialTheme && document.documentElement.classList.add('dark');
+  if(initialTheme) return('dark')
+  else return('light')
+};
```
Comment (krishnaacharyaa) [inline frontend/src/utils/theme-changer.ts:8]: please run `npm run format` in the frontend

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]

File: frontend/src/App.tsx (hunk 9 above)
Comment (Ritik-The-Dev) [reply frontend/src/App.tsx:40]: ok sir

File: frontend/src/utils/theme-changer.ts (hunk 10 above)
Comment (Ritik-The-Dev) [reply frontend/src/utils/theme-changer.ts:8]: ok sir

**Review by krishnaacharyaa — APPROVED** [review]
LGTM, Thank you for the PR

**Comment by krishnaacharyaa** [comment]
image

Can you please try rebasing it @Ritik-The-Dev 
If not we might have to create a new fresh PR with your desired changes

**Comment by Ritik-The-Dev** [comment]
Sir I have Created a New Pull Request for better Understanding
